elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,459 @@
+
+Within WildFly Core the general `ExpressionResolver` support will be updated to detect the availability of an expression resolver capability exposing
+a runtime API for expression resolution.
+
+Also within WildFly Core the Elytron subsystem will be updated with new resources as required, resources will also contain operations both for key management
+and operations for the encryption of clear text strings for use as expressions in the management model.
+
+Additionally all documentation lives within the WildFly project so updates will be applied there accordingly.
+
+=== Other Interested Projects
+
+
+=== Relevant Installation Types
+
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+As a core management feature this enhancement will be applicable to all installation types.
+
+== Requirements
+
+=== Hard Requirements
+
+The overall intent of this enhancement is for the Elytron subsystem to provide a configurable expression resolver resource to make use of an `AES`
+`SecretKey` loaded from a `CredentialStore` to decrypt expressions in the management model which have been previously encrypted using the same key.  To
+meet this requirement some additional functionality will also be required to hande the management of the keys.
+
+It is a requirement within this enhancement that multiple secret keys can be used for the decryption of expressions, these keys could be loaded from the same
+credential store or from separate credential stores - this relates to the "chicken and egg" problem which will be described later.
+
+The initial integration of this enhancement will not support specifying custom `java.security.Provider` instances as the key type and transformation will be
+using pre-defined defaults - however the expression resolver will be initialised after the Elytron subsystem has performed initialisation and registration of
+the providers.
+
+This enhancement will exclusively make use of `AES` `javax.crypt.SecretKey` instances, although it is possible to avoid the use of a
+`javax.crypto.SecretKeyFactory` as the keys can be generated using random key material directly using the `javax.crypto.spec.SecretKeySpec` class we should
+still use the `SecretKeyFactory` in case certified providers are added for key generation.  The generated `SecretKey` will be stored in a `CredentialStore`.
+
+==== Credential Store
+
+==== PropertiesCredentialStore / secret-key-credential-store
+
+A new credential store type will be implemented, this credential store will be dedicated to the storage of `SecretKeyCredential` instances which will be stored
+in a properties file.  Unlike other credential store types this implementation will not be adding any additional protection to the keys it contains, this
+credential store is primarily intended to solve the problem of how we provide the first secret key to the application server process without relying on a hard
```

Comment:
I have just realized the an external source can be used for the master password, which is not documented upstream, see https://issues.redhat.com/browse/WFLY-14589

For the documentation (analysis) purposes, what is the preferred way how to secure the master password? For example, I am not sure whether `secret-key-credential-store` could be used for a FIPS compliant configuration.

Reply:
Yes using CMD or EXT to obtain a password to unlock a credential store does remain an option so if a user wants they don't have to use the secret-key-credential-store - overall they have a few options now to decide the most appropriate approach for their initial secrets.